app/services/katello/event_daemon/runner.rb

```diff
@@ -0,0 +1,113 @@
+      @services = {}
+      @cache = ActiveSupport::Cache::MemoryStore.new
+
+      class << self
+        def initialize
+          FileUtils.touch(lock_file)
+        end
+
+        def settings
+          SETTINGS[:katello][:event_daemon]
+        end
+
+        def pid
+          return unless pid_file && File.exist?(pid_file)
+
+          File.open(pid_file) { |f| f.read.to_i }
+        end
+
+        def pid_file
+          pid_dir.join('katello_event_daemon.pid')
+        end
+
+        def pid_dir
+          Rails.root.join('tmp', 'pids')
+        end
+
+        def lock_file
+          Rails.root.join('tmp', 'katello_event_daemon.lock')
+        end
+
+        def write_pid_file
+          return unless pid_file
+
+          FileUtils.mkdir_p(pid_dir)
+          File.open(pid_file, 'w') { |f| f.puts Process.pid }
+        end
+
+        def stop
+          return unless pid == Process.pid
+          Rails.logger.info "<<<<<<<<<<<<<<< STOPPING RUNNER >>>>>>>>>>>>>>>>>"
+          @monitor_thread.kill
+          @cache.clear
+          @services.values.each(&:close)
+          File.unlink(pid_file) if pid_file && File.exist?(pid_file)
+        end
+
+        def start
+          Rails.logger.info "((((((((((((((NOT RUNNABLE))))))))))))))" unless runnable?
+          return unless runnable?
+          lockfile = File.open(lock_file, 'r')
+          begin
+            lockfile.flock(File::LOCK_EX)
+            strtd = started?
+            Rails.logger.info("******************** ABORTING start; already started *************") if strtd
+            return if strtd # ensure it wasn't started while we waited for the lock
+            start_monitor_thread
+            write_pid_file
+
+            at_exit do
+              stop
+            end
+
+            Rails.logger.info("Katello event daemon started process=#{Process.pid}")
+          ensure
+            lockfile.flock(File::LOCK_UN)
+          end
+        end
+
+        def started?
+          Process.kill(0, pid)
+          true
+        rescue Errno::ESRCH, TypeError # process no longer exists or we had no PID cached
+          false
+        end
+
+        def start_monitor_thread
+          @monitor_thread = Thread.new do
+            Katello::EventDaemon::Monitor.new(@services.values).start
+          end
+        end
+
+        def runnable?
+          # avoid accessing the disk on each request
+          @cache.fetch('katello_event_daemon_runnable', expires_in: 1.minute) do
+            !started? && settings[:enabled] && !::Foreman.in_rake? && !Rails.env.test?
+          end
+        end
+
+        def register_service(name, klass)
+          @services[name] = klass
+        end
+
+        def service_statuses
+          result = {}
+          @services.each do |service_name, service_class|
+            result[service_name] = service_class.status
+          end
+          result
+        end
+
+        def cached_status(service_name = nil)
```

Comment:
IMO this method could/should be renamed to `service_status`. The cache aspect is not relevant to this external API, although you and I both know where the status comes from!

that simplifies things significantly because this also removes the logic in the last line of the method which alters the return value (not a fan of that pattern personally):

```
def service_status(service_name)
  Rails.cache.read(STATUS_CACHE_KEY)&.dig(service_name)
end
```

This is unneeded now, but in the future if we wanted all services status, we just add another method and keep the two concepts separate:
```
def services_status
  Rails.cache.read(STATUS_CACHE_KEY)
end
```

Reply:
The future `services_status` will go great right next to `service_statuses` :laughing: